raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -107,53 +112,106 @@ private OptionalLong quorumMajorityFetchTimestamp() {
-                   .map(state -> state.nodeId).collect(Collectors.toList());
+                   .map(state -> state.nodeId)
+                   .collect(Collectors.toList());
     }
 
-    private List<ReplicaState> followersByDescendingFetchOffset() {
-        List<ReplicaState> followersByDescendingFetchOffset = new ArrayList<>(this.voterReplicaStates.values());
-        Collections.sort(followersByDescendingFetchOffset);
-        return followersByDescendingFetchOffset;
+    private List<VoterState> followersByDescendingFetchOffset() {
+        return new ArrayList<>(this.voterReplicaStates.values())
+            .stream().sorted().collect(Collectors.toList());
     }
 
     /**
      * @return true if the high watermark is updated too
      */
     public boolean updateEndOffset(int remoteNodeId, LogOffsetMetadata endOffsetMetadata) {
-        ReplicaState state = ensureValidVoter(remoteNodeId);
+        // Ignore fetches from negative replica id, as it indicates
+        // the fetch is from non-replica. For example, a consumer.
+        if (remoteNodeId < 0) {
+            return false;
+        }
+
+        ReplicaState state = getReplicaState(remoteNodeId);
+
         state.endOffset.ifPresent(currentEndOffset -> {
             if (currentEndOffset.offset > endOffsetMetadata.offset)
                 throw new IllegalArgumentException("Non-monotonic update to end offset for nodeId " + remoteNodeId);
         });
-        state.hasEndorsedLeader = true;
+
         state.endOffset = Optional.of(endOffsetMetadata);
-        return updateHighWatermark();
+
+        if (isVoter(remoteNodeId)) {
+            ((VoterState) state).hasEndorsedLeader = true;
+            addEndorsementFrom(remoteNodeId);
+            return updateHighWatermark();
+        }
+        return false;
     }
 
     public void addEndorsementFrom(int remoteNodeId) {
-        ReplicaState replicaState = ensureValidVoter(remoteNodeId);
-        replicaState.hasEndorsedLeader = true;
+        VoterState voterState = ensureValidVoter(remoteNodeId);
+        voterState.hasEndorsedLeader = true;
     }
 
-    private ReplicaState ensureValidVoter(int remoteNodeId) {
-        ReplicaState state = voterReplicaStates.get(remoteNodeId);
+    private VoterState ensureValidVoter(int remoteNodeId) {
+        VoterState state = voterReplicaStates.get(remoteNodeId);
         if (state == null)
             throw new IllegalArgumentException("Unexpected endorsement from non-voter " + remoteNodeId);
         return state;
     }
 
+    ReplicaState getReplicaState(int remoteNodeId) {
+        ReplicaState state = voterReplicaStates.get(remoteNodeId);
+        if (state == null) {
+            observerReplicaStates.putIfAbsent(remoteNodeId, new ReplicaState(remoteNodeId));
+            return observerReplicaStates.get(remoteNodeId);
+        }
+        return state;
+    }
+
+    List<DescribeQuorumResponseData.ReplicaState> getVoterStates() {
```

Comment:
Have the LeaderState depending on a `DescribeQuorumResponseData` is a bit too penetrating and we only need it as the return type. Could we move these three functions to KafkaRaftClient instead taking the `ReplicaStates` map as parameter? Also maybe these three can be static functions too?

Reply:
It's a bit tricky because if we take the conversion function out of `LeaderState`, we need to expose `ReplicaState/VoterState` to the `KafkaRaftClient` unnecessarily as well. The coupling is less with the current setup IMHO.

Reply:
Do we have to expose the `ReplicaState/VoterState`? I thought we can just expose a `Map<clientId, long>` as replicaId -> offset?

Reply:
Sg